gateway-sp-comms/src/communicator.rs

```diff
@@ -0,0 +1,462 @@
+            .request_response(
+                &controller,
+                request,
+                Some(timeout),
+                ResponseKindExt::try_into_bulk_ignition_state,
+            )
+            .await?;
+
+        // deserializing checks that `num_targets` is reasonably sized, so we
+        // don't need to guard that here
+        let targets =
+            &bulk_state.targets[..usize::from(bulk_state.num_targets)];
+
+        // map ignition target indices back to `SpIdentifier`s for our caller
+        targets
+            .iter()
+            .copied()
+            .enumerate()
+            .map(|(target, state)| {
+                let port = self
+                    .switch
+                    .switch_port_from_ignition_target(target)
+                    .ok_or(SpCommunicationError::BadIgnitionTarget(target))?;
+                let id = self.port_to_id(port);
+                Ok((id, state))
+            })
+            .collect()
+    }
+
+    /// Instruct the local ignition controller to perform the given `command` on
+    /// `target_sp`.
+    pub async fn send_ignition_command(
+        &self,
+        target_sp: SpIdentifier,
+        command: IgnitionCommand,
+        timeout: Instant,
+    ) -> Result<(), Error> {
+        let controller = self.switch.ignition_controller();
+        let target = self.id_to_port(target_sp)?.as_ignition_target();
+        let request = RequestKind::IgnitionCommand { target, command };
+
+        self.request_response(
+            &controller,
+            request,
+            Some(timeout),
+            ResponseKindExt::try_into_ignition_command_ack,
+        )
+        .await
+    }
+
+    /// Get our current serial console contents for the given SP component.
+    pub fn serial_console_contents(
+        &self,
+        sp: SpIdentifier,
+        component: &SpComponent,
+    ) -> Result<Option<SerialConsoleContents>, Error> {
+        let port = self.id_to_port(sp)?;
+        Ok(self.recv_handler.serial_console_contents(port, component))
+    }
+
+    /// Send `data` to the given SP component's serial console.
+    pub async fn serial_console_send(
+        &self,
+        sp: SpIdentifier,
+        component: &SpComponent,
+        data: &[u8],
+        timeout: Instant,
+    ) -> Result<(), Error> {
+        let port = self.id_to_port(sp)?;
+        let sp =
+            self.switch.sp_socket(port).ok_or(Error::SpAddressUnknown(sp))?;
+
+        // TODO how do we handle multiple serial console sends to the same SP at
+        // the same time? in a previous iteration of this code I had a mutex
+        // here, but that only protects against multiple posts going through the
+        // same MGS instance - we have two, and local mutexes won't help if both
+        // are trying to send to the same SP simultaneously. maybe we need some
+        // kind of time-limited lock from the SP side where it discards any
+        // incoming data from other sources?
```

Comment:
Note - this comment refers to a behavior change compared to what was in the old version of the code. Open to suggestions about what to do here; I suspect the relatively naive mutex solution I had gave the illusion of addressing a problem without really addressing it. It solves "two simultaneous POSTs to a single SP's console via the same MGS instance", but doesn't address simultaneous requests to two MGS instances, or any of the variety of failure modes (e.g., we time out in the middle of a stream of UDP packets).

Reply:
If we want to enforce mutual exclusion, the SP would need to play a role in it. One could imagine a few different mechanisms. For example, the SP could issue a console token (time-limited, as you say) that a client (presumably nexus or something else talking to one or either MGS) could use to permit console writes. Alternatively, the SP could issue a time-limited token to an MGS and clients would need to remember to route traffic through the proper MGS... and MGS would need to enforce mutual exclusion between clients.

I think any of that would be overkill. The serial console is really an interface of last resort. Until we've made some progress with RFD 247 (and we should! soon!), I'd argue that we should take the simplest path which is to provide no mechanism for mutual exclusion whatsoever. My guess is that 247 will land at a spot where that is sufficient for MVP